### PR TITLE
Arguments starting with "__" are now positional-only

### PR DESCRIPTION
**Pull request by sixolet**
1. When you are defining a function, mypy will now consider any argument starting with `__` to not have a name -- ie, be positional-only. 
2. It will also consider *all* arguments of dunder-functions other than `__new__` and `__init__` to be positional-only, as matches how they're expected to be used.

This PR is in preparation for a larger PR (to come, see also https://github.com/sixolet/mypy/tree/function-compatibility ) that rewrites the function compatibility rules.  Specifically, a function with a named positional argument will be allowed to be a subtype of a function with a corresponding unnamed positional argument, but will not be allowed to be a subtype of a function with a corresponding differently-named positional argument. 

Effect 1 of this change will allow writing function definitions that have all possible function types (except `Callable[..., blah]`).  For people designing superclasses carefully, they will be able to specifically notate that their functions take arguments whose names don't matter, and they expect never to be passed by name.  This gives subclass implementors the leeway to make their own decisions about argument names (which, they mostly happen to take anyway).

Effect 2 of this change will make the function compatibility PR much easier to land, as it'll avoid a certain awkward unioning of "forward operators" in stubs resulting in types that are unions of callable types with inconveniently incompatible names, which ends up not callable.

### CODE REVIEW

**Comment by sixolet** [comment]
Since it's hard to actually pull out effects of namelessness of arguments to functions before the function compatibility change lands, I've put some tests for it in that branch rather than this one.

I *can* test it in this PR by writing a test that attempts to call functions with arguments that start with double underscores by name, but since that'll be a place where the actual behavior of the interpreter differs from what I'm making mypy consider an error, I'm not sure that's the best plan.

File: mypy/parse.py [hunk 1]
```diff
@@ -819,8 +836,12 @@ def verify_argument_kinds(self, kinds: List[int], line: int, column: int) -> Non
...
             found.add(kind)
 
-    def construct_function_type(self, args: List[Argument], ret_type: Type,
-                                line: int, column: int) -> CallableType:
+    def construct_function_type(self,
+                                args: List[Argument],
+                                ret_type: Type,
+                                line: int,
+                                column: int,
+                                include_names: bool=True) -> CallableType:
```
Comment (JukkaL) [inline mypy/parse.py]: Add spaces around `=`.

File: mypy/parse.py [hunk 2]
```diff
@@ -14,6 +14,9 @@
     UnicodeLit, FloatLit, Op, Indent, Keyword, Punct, LexError, ComplexLit,
     EllipsisToken
 )
+
```
Comment (JukkaL) [inline mypy/parse.py]: Style nit: remove empty lines around import.

File: mypy/fastparse2.py [hunk 3]
```diff
@@ -18,6 +18,9 @@
 import sys
 
 from typing import Tuple, Union, TypeVar, Callable, Sequence, Optional, Any, cast, List
+
+from mypy.sharedparse import special_function_elide_names
```
Comment (JukkaL) [inline mypy/fastparse2.py]: Style nit: remove empty lines around import.

File: mypy/fastparse.py [hunk 4]
```diff
@@ -2,6 +2,9 @@
 import sys
 
 from typing import Tuple, Union, TypeVar, Callable, Sequence, Optional, Any, cast, List
+
+from mypy.sharedparse import special_function_elide_names
```
Comment (JukkaL) [inline mypy/fastparse.py]: Style nit: remove empty lines around import.

File: mypy/sharedparse.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+"""Shared logic between our three mypy parser files.
```
Comment (JukkaL) [inline mypy/sharedparse.py]: Style nit: In single-line docstrings, we keep both `"""`s on the same line (e.g. `"""Doc string."""`).

**Comment by JukkaL** [comment]
> I can test it in this PR by writing a test that attempts to call functions with arguments that start with double underscores by name, but since that'll be a place where the actual behavior of the interpreter differs from what I'm making mypy consider an error, I'm not sure that's the best plan.

I'd like to have tests like these. For example, one that tries to call a `__eq__` method with a keyword argument, and another that tries to call a regular function (and a method) using keyword arguments with names prefixed with `__`. It doesn't matter if mypy differs from Python semantics -- we still want to test that mypy behaves as expected. We don't expect to be able to run unit tests as Python programs.

Finally, I wonder if we should only handle known dunder methods using a whitelist. I can define my own method named `__my_init__` and expect to be able to call it with keyword arguments. This may be questionable stylistically, but it still works, and I'd rather not break code like that.

**Comment by sixolet** [comment]
Ok, I think that's all your suggestions, @JukkaL

**Comment by JukkaL** [comment]
This crashes when run against the 'C' repo. Traceback extract:

```
  File "/Users/jukka/src/test-mypy-pr/mypy/mypy/fastparse2.py", line 154, in translate_stmt_list
    stmt = self.visit(e)
  File "/Users/jukka/src/test-mypy-pr/server/.mypy/venv/lib/python3.5/site-packages/typed_ast/ast27.py", line 241, in visit
    return visitor(node)
  File "/Users/jukka/src/test-mypy-pr/mypy/mypy/fastparse2.py", line 112, in wrapper
    node = f(self, ast)
  File "/Users/jukka/src/test-mypy-pr/mypy/mypy/fastparse2.py", line 275, in visit_FunctionDef
    arg_names = [None if name.startswith("__") else name for name in arg_names]
  File "/Users/jukka/src/test-mypy-pr/mypy/mypy/fastparse2.py", line 275, in <listcomp>
    arg_names = [None if name.startswith("__") else name for name in arg_names]
AttributeError: 'Tuple' object has no attribute 'startswith'
```

This seems to be caused by tuples in function headers (Python 2 only). For example:
```
def f(x, (y, z)): ...
```

**Comment by sixolet** [comment]
Fun.  Where is this 'C' repo to test against, and also if I write `# flags: --python-version 2.7` in a test with and without `--fast-parser` will it do the right thing?

File: mypy/sharedparse.py [hunk 6]
```diff
@@ -90,3 +92,7 @@
 
 def special_function_elide_names(name: str) -> bool:
     return name in MAGIC_METHODS_POS_ARGS_ONLY
+
+
+def argument_elide_name(name: Union[str, Tuple, None]) -> bool:
```
Comment (JukkaL) [inline mypy/sharedparse.py]: The type of the `name` argument is funny. It seems that the fast parser does some unsafe things and doesn't always conform to type annotations. Anyway, this is an issue with the fast parser only.

**Pull request merged**